### DISCUSSION TITLE
Implementing an equaliser example

### DISCUSSION THREAD

**Discussion by AvishkaDon** [pschatzmann/arduino-audio-tools#124] [Q&A] [answered]
Hi,
I am using your a2dp library as sink on my esp32 to send data from a Bluetooth device and the esp32 connected to a I2S DAC. I would like do some media processing on esp32 before outputting the data to the DAC (I.e an [equaliser](https://espressif-docs.readthedocs-hosted.com/projects/esp-adf/en/latest/api-reference/audio-processing/equalizer.html) or some kind of a bass boost function). I was wondering if it’s possible to use arduino audio tools library do this? If so do you have any examples?

Many thanks 🙏

**Comment by pschatzmann** [comment]
Unfortunately I do not provide any easy to use Equalizer. Though it would be possible to build one using Filters, I guess that might be a bit challenging to calculate the input parameters.

**Reply by AvishkaDon** [reply]
I was going through some of the examples on the filters and I found [stream-i2s-filter-i2s](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-filter-i2s/streams-i2s-filter-i2s.ino)
Is there a way to get the Bluetooth sink stream as `I2SStream in`? or is it possible to pass in [`[A2DPStream]`](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-a2dp/streams-i2s-a2dp.ino) to `FilteredStream`?

**Comment by pschatzmann** [comment]
Yes, that should work. Or if you use [this example](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api/basic-a2dp-i2s) you can wrap the I2SStream in a FilteredStream.

**Reply by AvishkaDon** [reply]
oh cool! - did you mean something like this? :) `/**
 * @file basic-a2dp-audioi2s.ino
 * @brief A2DP Sink with output to I2SStream. This example is of small value
 * since my Bluetooth Library already provides I2S output out of the box.
 *
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#define USE_A2DP
// #include "AudioConfigLocal.h"
#include "AudioTools.h"

uint16_t sample_rate=44100;
uint16_t channels = 2;
BluetoothA2DPSink a2dp_sink;
I2SStream i2s_in;          
I2SStream i2s_out;          



// copy filtered values
FilteredStream<int16_t, float> filtered(i2s_in, channels);  // Defiles the filter as BaseConverter
StreamCopy copier(i2s_out, filtered);// copies sound into i2s

float coef[] = { 0.021, 0.096, 0.146, 0.096, 0.021}; //these values will change for  a low-pass!


// Write data to SPDIF in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s_in.write(data, length); //input data to the filter
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  filtered.setFilter(0, new FIR<float>(coef));
  filtered.setFilter(1, new FIR<float>(coef));

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);

  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(true);
  a2dp_sink.start("a2dp-i2s");

  //i2s_out internal DAC
  auto config = i2s_out.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  i2s_out.begin(config);


  //i2s_out to expternal DAC
  auto config_out = i2s_out.defaultConfig(TX_MODE);
  config_out.sample_rate = sample_rate; 
  config_out.bits_per_sample = 16;
  config_out.i2s_format = I2S_STD_FORMAT;
  config_out.is_master = true;
  config_out.port_no = 1;
  config_out.pin_ws = 17;
  config_out.pin_bck = 18;
  config_out.pin_data = 19;
  i2s_out.begin(config_out);
  config_out.fixed_mclk = sample_rate * 256;
  config_out.pin_mck = 0;

  Serial.println("I2S started...");
}`

**Comment by pschatzmann** [comment]
No, it would be something as follows:
```
BluetoothA2DPSink a2dp_sink;
I2SStream i2s_out;
FilteredStream<int16_t, float> filtered(i2s_out, channels);  // filter on output
// copy and filter values
StreamCopy copier(filtered, a2dp_sink);// copies sound into i2s
```

Or if you use [this example](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api/basic-a2dp-i2s) you can omit the StreamCopy and write to filtered inesead of i2s

**Reply by AvishkaDon** [reply]
Is this what you meant by bypassing the `StreamCopy` with the "[basic-a2dp-i2s](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api/basic-a2dp-i2s)" example? 
Also, I am writing to the internal DAC of the esp32 and I am don't think I am doing this correctly as I am not getting any signals out of the pin 25 and 26.
```
#define USE_A2DP
#include "AudioTools.h"

uint16_t sample_rate=44100;
uint16_t channels = 2;

BluetoothA2DPSink a2dp_sink;
I2SStream i2s;
FilteredStream<int16_t, float> filtered(i2s, channels);  // filter on output
float coef[] = { 0.021, 0.096, 0.146, 0.096, 0.021}; //these values will change for  a low-pass!


// Write data to SPDIF in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s.write(data, length); //input data to the filter
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  filtered.setFilter(0, new FIR<float>(coef));
  filtered.setFilter(1, new FIR<float>(coef));

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);

  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(true);
  a2dp_sink.start("a2dp-i2s");

  //i2s_out esp32 internal DAC
  auto config = i2s.defaultConfig(TX_MODE);
  config.sample_rate = a2dp_sink.sample_rate();
  config.channels = 2;
  config.sample_rate = sample_rate; 
  config.bits_per_sample = 16;
  i2s.begin(config);
}

void loop() { }
```

**Comment by pschatzmann** [comment, marked as answer]
Two points: 
- If you want to use the internal DAC you need to replace I2SStream with AnalogAudioStream.
- If you want to apply the filter you need to replace i2s.write() with filtered.write() - otherwise you just write directly to i2s w/o giving the filter a chance to do anything...

**Reply by AvishkaDon** [reply]
Many thanks! 🙏  this helped out quite a lot

**Comment by pschatzmann** [comment]
In the meantime I have added a draft version of [Equilizer3Bands](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_equilizer3_bands.html).

**Comment by daviti231** [comment]
thank you very much for your reply, sorry if I trouble you.

**Comment by burnutec** [comment]
Hey , it may not be relatable to this , but would it be possible to use this library with the pcm5101/5102 dac ? Since that has a better solderability to do as a DIY -er. 
Thanks, and sorry if it is completely irrelevant. :)

**Comment by pschatzmann** [comment]
Just read the [external DAC documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC)
